File: 2025/2025-01.c
// clang -std=c23 -I../libb/src ../libb/src/*.c -o 2025-01 2025-01.c &&
// ./2025-01

#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <time.h>

#include "libb.h"

typedef Vec(i64) Turns;

void print_turns(Turns turns) {
  for (size_t i = 0; i < turns.n; i++) {
    printf("%ld\n", turns.p[i]);
  }
}

Turns parse(Arena *a, Str input) {
  Turns v = {};
  vec_init(&v, a, 200);
  auto it = str_split_lines(input);
  Str s;
  while (str_iter_split_next(&it, &s)) {
    if (s.n == 0)
      break;

    auto res = str_to_i64(str_slice(s, 1, s.n - 1));
    assert(res.ok);
    i64 n = res.val;
    if (s.p[0] == 'L') {
      n *= -1;
    }
    vec_push(&v, n);
  }
  return v;
}

i64 part_1(Arena *a, Str input) {
  auto turns = parse(a, input);
  i64 pos = 50;
  i64 cross_counts = 0;
  for (size_t i = 0; i < turns.n; i++) {
    i64 turn = turns.p[i];
    pos = (pos + turn) % 100;
    if (pos == 0) {
      cross_counts += 1;
    }
  }

  return cross_counts;
}

i64 part_2(Arena *a, Str input) {
  auto turns = parse(a, input);

  i64 pos = 50;
  i64 cross_counts = 0;

  for (size_t i = 0; i < turns.n; i++) {
    i64 turn = turns.p[i];

    auto q = div(pos + turn, 100);
    auto n_turns = q.quot;
    auto next_pos = q.rem;
    if (next_pos < 0) {
      next_pos = 100 + next_pos;
      n_turns -= 1;
    }

    if (turn >= 0) {
      cross_counts += n_turns;
      pos = next_pos;
    } else {
      cross_counts += abs(n_turns);
      if (next_pos == 0) {
        cross_counts += 1;
      }
      if (pos == 0) {
        cross_counts -= 1;
      }
      pos = next_pos;
    }
  }
  return cross_counts;
}

int main(void) {
  Arena a = {};
  arena_init(&a, 10 * 1024 * 1024);
  FReadResult res = fs_read_file(&a, "./2025-01.input");
  if (!res.ok) {
    printf("Failed to read file: %u\n", res.err);
    return 1;
  };
  assert(res.ok);
  auto content = res.val;

  auto start = now_ns();
  auto p1 = part_1(&a, content);
  printf("2025\t1\t1\t%ld\t%ld\n", p1, (now_ns() - start));

  start = now_ns();
  auto p2 = part_2(&a, content);
  printf("2025\t1\t2\t%ld\t%ld\n", p2, (now_ns() - start));
}

File: 2025/2025-01.py
#!/usr/bin/env python

import time


def parse(input: list[str]) -> list[int]:
    turns: list[int] = []
    for line in input:
        i = int(line[1:])
        turns.append(i if line[0] == "R" else -i)
    return turns


def part_1(input):
    pos = 50
    cross_counts = 0
    turns = parse(input)
    for turn in turns:
        pos = (pos + turn) % 100
        if pos == 0:
            cross_counts += 1
    return cross_counts


def crosses_and_pos(start: int, turn: int) -> tuple[int, int]:
    turns, next_pos = divmod(start + turn, 100)
    if turn >= 0:
        return turns, next_pos
    # turn < 0
    turns = abs(turns)
    if next_pos == 0:
        turns += 1
    if start == 0:
        turns -= 1
    return turns, next_pos


def part_2(input):
    turns = parse(input)
    pos = 50
    cross_counts = 0
    for turn in turns:
        crosses, pos = crosses_and_pos(pos, turn)
        cross_counts += crosses
    return cross_counts


def _test():
    def assert_eq(a, b):
        assert a == b, f"{a} != {b}"

    assert_eq(crosses_and_pos(50, 49), (0, 99))
    assert_eq(crosses_and_pos(50, 50), (1, 0))
    assert_eq(crosses_and_pos(50, -50), (1, 0))
    assert_eq(crosses_and_pos(90, 99), (1, 89))
    assert_eq(crosses_and_pos(90, 100), (1, 90))
    assert_eq(crosses_and_pos(0, 100), (1, 0))
    assert_eq(crosses_and_pos(20, -20), (1, 0))
    assert_eq(crosses_and_pos(20, -120), (2, 0))
    assert_eq(crosses_and_pos(0, -99), (0, 1))
    assert_eq(crosses_and_pos(0, -100), (1, 0))
    assert_eq(crosses_and_pos(0, -101), (1, 99))
    assert_eq(crosses_and_pos(0, -200), (2, 0))
    assert_eq(crosses_and_pos(0, -652), (6, 48))


def run(fn, year=2025, day=1, part=0):
    start = time.perf_counter_ns()
    res = fn()
    elapsed_ns = time.perf_counter_ns() - start
    print(f"{year}\t{day}\t{part}\t{res}\t{elapsed_ns}")


if __name__ == "__main__":
    with open("2025-01.input") as f:
        input_file = [line.strip() for line in f.readlines()]
    _test()
    run(lambda: part_1(input_file), part=1)
    run(lambda: part_2(input_file), part=2)
